packages/serverless-components/nextjs-cdk-construct/src/index.ts

```diff
@@ -0,0 +1,376 @@
+            : {}),
+          [this.pathPattern("_next/data/*")]: {
+            viewerProtocolPolicy:
+              cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+            origin: new origins.S3Origin(bucket),
+            allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
+            cachedMethods: cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
+            compress: true,
+            cachePolicy: nextLambdaCachePolicy,
+            edgeLambdas
+          },
+          [this.pathPattern("_next/*")]: {
+            viewerProtocolPolicy:
+              cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+            origin: new origins.S3Origin(bucket),
+            allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
+            cachedMethods: cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
+            compress: true,
+            cachePolicy: nextStaticsCachePolicy
+          },
+          [this.pathPattern("static/*")]: {
+            viewerProtocolPolicy:
+              cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+            origin: new origins.S3Origin(bucket),
+            allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
+            cachedMethods: cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
+            compress: true,
+            cachePolicy: nextStaticsCachePolicy
+          },
+          ...(nextApiLambda
+            ? {
+                [this.pathPattern("api/*")]: {
+                  viewerProtocolPolicy:
+                    cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+                  origin: new origins.S3Origin(bucket),
+                  allowedMethods:
+                    cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
+                  cachedMethods:
+                    cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
+                  compress: true,
+                  cachePolicy: nextLambdaCachePolicy,
+                  edgeLambdas: [
+                    {
+                      eventType: cloudfront.LambdaEdgeEventType.ORIGIN_REQUEST,
+                      functionVersion: nextApiLambda.currentVersion
+                    }
+                  ]
+                }
+              }
+            : {})
+        }
+      }
+    );
+
+    const staticAssets = this.getStaticAssetsFromBuild();
+    Object.keys(staticAssets).forEach((staticAssetKey) => {
+      const staticAsset =
+        staticAssets[staticAssetKey as keyof typeof staticAssets];
+      const assetPath = path.join(
+        this.props.serverlessBuildOutDir,
+        "assets",
+        staticAsset
+      );
+      if (!fs.existsSync(assetPath)) return;
+      new s3Deploy.BucketDeployment(this, `AssetDeployment_${staticAssetKey}`, {
+        destinationBucket: bucket,
+        distribution: this.distribution,
+        destinationKeyPrefix: staticAsset,
+        distributionPaths: ["/*"],
+        sources: [s3Deploy.Source.asset(assetPath)],
+        cacheControl: [
```

Comment:
Pages would need different cache control here: `public, max-age=0, s-maxage=2678400, must-revalidate`.

And also since versioning pages, we are clearing old versions of pages with the latest deployment code (1.19 alpha) (keeping only last two deployments) to save S3 space

Reply:
I will have a look into how static pages are being versioned, but I presume this means no invalidation is required for these either which is nice! This is something I wanted to revisit as it's currently creating an invalidation on all resources every deploy which is unnecessary given everything is versioned?

Reply:
> And also in the latest 1.19 alpha, we are now versioning pages so only the last two deployments are kept to save S3 space.

I believe this is actually [handled by CDK itself](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-s3-deployment-readme.html#prune) as it automatically 'syncs' source assets, deleting ones that no longer exist. I'm assuming the reason you are keeping the last 2 deployments is for safety and thus this behavior is not needed for the construct?

Reply:
@dphang - update on this [here](https://github.com/serverless-nextjs/serverless-next.js/pull/878/commits/ed731a2e90c7427cbfe2287612d3be7d6914ebc1). I've refactored the `uploadStaticAssetsFromBuild` which was determining the cache-control for files so that it can be used in the CDK construct.